File: sort.cpp
#include <vector>
#include <iostream>

void printArray(const std::vector<int> &);
void insertionSort(std::vector<int> &);
void insertionSortImproved(std::vector<int> &);
std::vector<int> mergeSortedArray(std::vector<int> &, std::vector<int> &);
void mergeSort(std::vector<int> &);
void heapify(std::vector<int> &, int, int);
void buildHeap(std::vector<int> &);
void heapSort(std::vector<int> &);
int partition(std::vector<int> &, int, int);
void quickSort(std::vector<int> &, int, int);

int main()
{
    size_t vectorSize;    
    while (true)
    {        
        std::cin >> vectorSize;
        std::vector<int> vectorToSort(vectorSize);
        if (vectorSize == 0) break;
        for (int i = 0; i < vectorSize; i++)
        {
            std::cin >> vectorToSort[i];
        }
        //insertionSort(vectorToSort);
        //mergeSort(vectorToSort);
        heapSort(vectorToSort);
        //quickSort(vectorToSort, 0, vectorSize - 1);
        printArray(vectorToSort);
    }            
}

void insertionSort(std::vector<int> & aVector)
{
    int temp;
    for (int i = 0; i < aVector.size(); i++)
    {
        for (int j = i + 1; j < aVector.size(); j++)
        {
            if (aVector[i] > aVector[j])
            {
                temp = aVector[i];
                aVector[i] = aVector[j];
                aVector[j] = temp;
            }
        }
    }
}

void insertionSortImproved(std::vector<int> &aVector)
{
    bool changes;
    int temp;
    for (int i = 0; i < aVector.size(); i++)
    {
        changes = false;
        for (int j = i + 1; j < aVector.size(); j++)
        {
            if (aVector[i] > aVector[j])
            {
                temp = aVector[i];
                aVector[i] = aVector[j];
                aVector[j] = temp;
                changes = true;
            }
        }
        if (!changes) break;
    }
}

std::vector<int> mergeSortedArray(std::vector<int> &aVector1, std::vector<int> &aVector2)
{
    std::vector<int> mergedVector(aVector1.size() + aVector2.size());
    int i = 0;
    int j = 0;
    for (int k = 0; k < mergedVector.size(); k++)
    {
        if (j >= aVector2.size())
        {
            mergedVector[k] = aVector1[i];
            i++;
        } 
        else if (i >= aVector1.size())
        {
            mergedVector[k] = aVector2[j];
            j++;
        }
        else if (aVector1[i] <= aVector2[j])
        {
            mergedVector[k] = aVector1[i];
            i++;
        }
        else
        {
            mergedVector[k] = aVector2[j];
            j++;
        }                
    }
    return mergedVector;
}

void mergeSort(std::vector<int> & aVector)
{    
    if (aVector.size() > 1)
    {
        int middle = aVector.size() / 2;
        std::vector<int> left(aVector.begin(), aVector.begin() + middle);
        std::vector<int> right(aVector.begin() + middle, aVector.end());        
        mergeSort(left);
        mergeSort(right); 
        aVector = mergeSortedArray(left, right);
    }    
}

void heapify(std::vector<int> & aVector, int aPos, int aHeapSize)
{
    int temp;
    if (aVector.size() > 1)
    {
        int left = (2 * aPos) + 1;
        int right = left + 1;
        int largest = aPos;
        if (left < aHeapSize && aVector[left] > aVector[largest])
            largest = left;
        if (right < aHeapSize && aVector[right] > aVector[largest])
            largest = right;
        if (largest != aPos)
        {
            temp = aVector[aPos];
            aVector[aPos] = aVector[largest];
            aVector[largest] = temp;
            heapify(aVector, largest, aHeapSize);
        }
    }
}

void buildHeap(std::vector<int> & aVector)
{
    size_t vectorSize = aVector.size();
    for (int i = vectorSize / 2; i >= 0; i--)
        heapify(aVector, i, vectorSize);
}

void heapSort(std::vector<int> & aVector)
{
    buildHeap(aVector);
    for (int i = aVector.size() - 1; i >= 0; i--)
    {
        if (aVector[0] > aVector[i])
        {
            int temp = aVector[0];
            aVector[0] = aVector[i];
            aVector[i] = temp;
            heapify(aVector, 0, i - 1);
        }
    }
}

int partition(std::vector<int> & aVector, int aLeft, int aRight)
{    
    int pivot = aVector[(aLeft + aRight) / 2];
    //std::cout << "pivot = " << pivot << std::endl;
    int i = aLeft;
    int j = aRight;
    while (true)
    {
        while (aVector[j] > pivot) j--;        
        while (aVector[i] < pivot) i++;         
        if (i <= j)
        {
            int temp = aVector[i];
            aVector[i] = aVector[j];
            aVector[j] = temp;
            i++;
            j--;
        } 
        else return j;
    }
}

void quickSort(std::vector<int> & aVector, int aLeft, int aRight)
{    
    if (aLeft < aRight)
    {
        int middle = partition(aVector, aLeft, aRight);  
        if (aLeft < middle)        
            quickSort(aVector, aLeft, middle);
        if (middle < aRight)
            quickSort(aVector, middle + 1, aRight);
    }
}

void printArray(const std::vector<int> & aVector)
{
    for (int item : aVector)
    {
        std::cout << item << " ";
    }
    std::cout << std::endl;        
}

File: Exercicio_A/sort.py

import sys
import random


def print_array(alist):
    """Print an array without brackets and commas."""
    print(*alist, sep=' ')


def insertionsort(alist):
    """Sorting an array using insertionsort algorithm."""
    for i in range(0, len(alist)):
        for j in range(i + 1, len(alist)):
            if alist[i] > alist[j]:
                alist[i], alist[j] = alist[j], alist[i]
    # return alist


def insertionsort_improved(alist):
    """Sorting an array using insertionsort algorithm with improvement."""
    for i in range(len(alist), 0, -1):
        changes = False
        for j in range(0, i-1):
            if alist[j] > alist[j + 1]:
                alist[j], alist[j + 1] = alist[j + 1], alist[j]
                changes = True

        if not changes:
            break
    # return alist


def merge_sorted_array(alist1, alist2):
    """Merge two ordered arrays."""
    i, j = 0, 0
    N = len(alist1) + len(alist2)
    merged_list = [None] * N
    for k in range(0, N):
        if j > (len(alist2) - 1):
            merged_list[k] = alist1[i]
            i += 1
        elif i > (len(alist1) - 1):
            merged_list[k] = alist2[j]
            j += 1
        elif alist1[i] <= alist2[j]:
            merged_list[k] = alist1[i]
            i += 1
        else:
            merged_list[k] = alist2[j]
            j += 1
    return merged_list


def mergesort(alist):
    """Sorting an array using mergesort algorithm."""
    if len(alist) > 1:
        middle = len(alist) // 2
        left = mergesort(alist[:middle])
        right = mergesort(alist[middle:])
        alist = merge_sorted_array(left, right)
    return alist


def heapify(alist, i, heap_size):
    if len(alist) > 1:
        left = (2 * i) + 1
        right = (2 * i) + 2
        largest = i
        if left < heap_size and alist[left] > alist[largest]:
            largest = left
        if right < heap_size and alist[right] > alist[largest]:
            largest = right
        if largest != i:
            alist[i], alist[largest] = alist[largest], alist[i]
            heapify(alist, largest, heap_size)


def build_heap(alist):
    """Build a heap from an array."""
    for i in range((len(alist) // 2), -1, -1):
        heapify(alist, i, len(alist))


def heapsort(alist):
    """Sorting an array using heapsort algorithm."""
    if len(alist) > 1:
        build_heap(alist)
        for i in range(len(alist)-1, 0, -1):
            if alist[0] > alist[i]:
                alist[0], alist[i] = alist[i], alist[0]
                heapify(alist, 0, i - 1)
    # return alist


def partition(alist, aleft, aright):
    pivot = alist[random.randint(aleft, aright)]    
    i = aleft - 1
    j = aright + 1
    while True:        
        while True:
            j -= 1
            if alist[j] <= pivot:
                break            
        while True:
            i += 1            
            if alist[i] >= pivot:
                break            
        if i < j:
            alist[i], alist[j] = alist[j], alist[i]
        else:
            return j


def quicksort(alist, aleft, aright):
    """Sorting an array using quicksort algorithm."""
    if aleft < aright:
        middle = partition(alist, aleft, aright)
        quicksort(alist, aleft, middle)
        quicksort(alist, middle + 1, aright)

while True:
    N = sys.stdin.readline()
    if N.strip() == '0':        
        break
    else:        
        V = [int(i) for i in sys.stdin.readline().split()]
        #insertionsort(V)
        #insertionsort_improved(V)
        #V = mergesort(V)
        #quicksort(V, 0, len(V) - 1)        
        heapsort(V)
        print_array(V)
